library/constant_time.c

```diff
@@ -47,16 +47,58 @@
 
 #include <string.h>
 
+/*
+ * Define MBEDTLS_EFFICIENT_UNALIGNED_VOLATILE_ACCESS where assembly is present to
+ * perform fast unaligned access to volatile data.
+ *
+ * This is needed because mbedtls_get_unaligned_uintXX etc don't support volatile
+ * memory accesses.
+ *
+ * Some of these definitions could be moved into alignment.h but for now they are
+ * only used here.
+ */
+#if defined(MBEDTLS_EFFICIENT_UNALIGNED_ACCESS) && defined(MBEDTLS_HAVE_ASM)
+#if defined(__arm__) || defined(__thumb__) || defined(__thumb2__) || defined(__aarch64__)
+#define MBEDTLS_EFFICIENT_UNALIGNED_VOLATILE_ACCESS
+#endif
+#endif
+
+#if defined(MBEDTLS_EFFICIENT_UNALIGNED_VOLATILE_ACCESS)
+static inline uint32_t mbedtls_get_unaligned_volatile_uint32( volatile const unsigned char *p )
+{
+    /* This is UB, even where it's safe:
+    *    return *((volatile uint32_t*)p);
+    * so instead the same thing is expressed in assembly below.
+    */
+    uint32_t r;
+#if defined(__arm__) || defined(__thumb__) || defined(__thumb2__)
+    asm( "ldr %0, [%1]" : "=r"(r) : "r"(p) : );
+#elif defined(__aarch64__)
+    asm( "ldr %w0, [%1]" : "=r"(r) : "r"(p) : );
+#endif
+    return r;
+}
+#endif /* MBEDTLS_EFFICIENT_UNALIGNED_VOLATILE_ACCESS */
+
 int mbedtls_ct_memcmp( const void *a,
                        const void *b,
                        size_t n )
 {
-    size_t i;
+    size_t i = 0;
     volatile const unsigned char *A = (volatile const unsigned char *) a;
     volatile const unsigned char *B = (volatile const unsigned char *) b;
-    volatile unsigned char diff = 0;
+    volatile uint32_t diff = 0;
```

Comment:
I know it was here previously, but I don't think `volatile` is necessary or helpful here.

Looking at compiler output, having `volatile` causes an extra store to memory in the stack frame; not having it doesn't seem to cause any warnings

Reply:
Looks like @gabor-mezei-arm added the `volatile` in db9a38c672 - do you know the reason for this Gabor?

Reply:
Gábor was moving some preexisting code (formerly in `include/mbedtls/ssl_internal.h`). The `volatile` keyword for `diff` was added by Hanno in 59e6963a37a615b137ee4f9824f798dc704fdd96 via https://github.com/Mbed-TLS/mbedtls-restricted/pull/370.

Reply:
There are comments in that commit message that should really be comments in the code!

Reply:
I moved the different  `memcmp` implementations and I kept the 'volatile' version from them. I don't remember the reason.

Reply:
Interesting, it wasn't in every implementation (there were a few impls of `memcmp` in various files which were unified in `constant_time.c`). E.g., `cipher.c` and `rsa.c` did not have it, but `nist_kw.c` did.

Reply:
I would prefer to remove it, unless anyone is able to suggest a reason that it might be necessary?

Reply:
The divergence is because we had multiple copies and they didn't get updated at the same time (which is part of why we eventually unified them). When `volatile` was added to the SSL function in https://github.com/Mbed-TLS/mbedtls-restricted/pull/370, there was no other constant time memcmp. There was another copy in https://github.com/Mbed-TLS/mbedtls/pull/485 but it wasn't merged yet (that eventually happened via https://github.com/Mbed-TLS/mbedtls/pull/1617).

I am against removing `volatile` unless you find a flaw in the reasoning that led to adding it.

Reply:
From a quick read of #370: "theoretically opening the possibility for the compiler to notice that the loop operation diff |= A[i] ^ B[i] is pointless if diff = -1 (which happens very quickly if A and B are random and independent!)" - how would the compiler be able to determine this, given that A and B are volatile and therefore the compiler can't know their values at compile time?

Reply:
> A and B are volatile and therefore the compiler can't know their values at compile time

This does not follow. `volatile` mandates that the read takes place, but not that the compiler can't make assumptions on the value that is read. For that you need to manage to pass the pointer to something that the compiler can't analyze.

Compilers generally don't try to optimize when `^` is involved (because it doesn't appear in “normal” code, so compiler writers don't bother). So this code is likely safe in practice even without `volatile`. But in principle, they can.

Would `volatile register` help to avoid a memory access here? I'm not sure modern compilers pay any attention to `register`.

Reply:
A and B are pointers-to-volatile, so the compiler can't optimise those reads away. So AIUI the compiler will always have to execute the full loop.

Reply:
If we keep `volatile` in I would really like us to add the reasoning to the code - it would have saved all of this discussion had it been there

Reply:
OpenSSL doesn't have the `volatile` on its equivalent of `diff`: https://github.com/openssl/openssl/blob/master/crypto/cpuid.c#L184

Reply:
I discussed with Gilles & we agreed that there's no meaningful risk from dropping the volatile.

Reply:
If there is no further objection I will drop `volatile` from the definition of `diff`.

Reply:
Yes please. And while I think it's obvious, should we comment on the `volatile` for `A` and `B`?

Reply:
Consider the old code (because there's a lot less to follow), with a non-volatile `diff`:

```
    unsigned char diff = 0;
    for( i = 0; i < n; i++ ) {
        unsigned char x = A[i], y = B[i];
        diff |= x ^ y;
    }
    return( (int)diff );
```

A compiler is allowed to compile this as:
* while `i < n` and `x = A[i]; y = B[i]; diff != 0xff`: `diff |= A[i] ^ B[i]`
* while `i < n`: `x = A[i]; y = B[i];`
* return `diff`

Making `diff` volatile avoids this particular “optimization”.

I doubt that any existing compiler would look for such an “optimization” (which wouldn't make the code faster or smaller), but it would be allowed.